delocate/libsana.py

```diff
@@ -48,54 +252,131 @@ def tree_libs(start_path, filt_func=None):
-                    lib_path = resolve_rpath(install_name, search_paths)
-                elif install_name.startswith('@'):
-                    lib_path = install_name
-                else:
-                    lib_path = search_environment_for_lib(install_name)
-                if lib_path in lib_dict:
-                    lib_dict[lib_path][depending_libpath] = install_name
-                else:
-                    lib_dict[lib_path] = {depending_libpath: install_name}
+            for dependency_path, install_name in get_dependencies(
+                depending_path
+            ):
+                if dependency_path is None:
+                    # Mimic deprecated behavior.
+                    # A lib_dict with unresolved paths is unsuitable for
+                    # delocating, this is a missing dependency.
+                    dependency_path = realpath(install_name)
+                if install_name.startswith("@loader_path/"):
+                    # Support for `@loader_path` would break existing callers.
+                    logger.debug(
+                        "Excluding %s because it has '@loader_path'.",
+                        install_name
+                    )
+                    continue
+                lib_dict.setdefault(dependency_path, {})
+                lib_dict[dependency_path][depending_path] = install_name
     return lib_dict
 
 
+def resolve_dynamic_paths(lib_path, rpaths, loader_path, executable_path=None):
+    # type: (Text, Iterable[Text], Optional[Text], Optional[Text]) -> Text
+    """ Return `lib_path` with any special runtime linking names resolved.
+
+    If `lib_path` has `@rpath` then returns the first `rpaths`/`lib_path`
+    combination found.  If the library can't be found in `rpaths` then
+    DependencyNotFound is raised.
+
+    `@loader_path` and `@executable_path` are resolved with their respective
+    parameters.
+
+    Parameters
+    ----------
+    lib_path : str
+        The path to a library file, which may or may not be a relative path
+        starting with `@rpath`, `@loader_path`, or `@executable_path`.
+    rpaths : sequence of str
+        A sequence of search paths, usually gotten from a call to `get_rpaths`.
+    loader_path : str
+        The path to be used for `@loader_path`.
+        This should be the directory of the library which is loading `lib_path`.
+    executable_path : None or str, optional
+        The path to be used for `@executable_path`.
+        If None is given then the path of the Python executable will be used.
+
+    Returns
+    -------
+    lib_path : str
+        A str with the resolved libraries realpath.
+
+    Raises
+    ------
+    DependencyNotFound
+        When `lib_path` has `@rpath` in it but no library can be found on any
+        of the provided `rpaths`.
+    """
+    if executable_path is None:
+        executable_path = dirname(sys.executable)
+    if lib_path.startswith('@loader_path/'):
+        if loader_path is None:
```

Comment:
`loader_path` still allowed to be None?

Reply:
My mistake.  I've removed the Optional type and this check.